File: Companion/Riff.Core/SoundPacks.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace Riff.Core;

public record PackSound(string Id, string Name, string FileName, string SourceURL, string Provider,
    string Uploader, string Rights, int ByteCount, string Sha256, double Duration)
{
    public string? OriginalDownloadURL { get; init; }
    public string? Acquisition { get; init; }
    public string ClipId => "pack-" + Id;
    public void Validate(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length != ByteCount || !Convert.ToHexString(SHA256.HashData(bytes)).Equals(Sha256, StringComparison.OrdinalIgnoreCase))
            throw new ArgumentException($"The audio for {Name} changed or is incomplete. Update Riff and try again.");
    }
}
public record SoundPack(string Id, string Name, string Description, string Icon, string Color, List<PackSound> Sounds);

public static class SoundPacks
{
    public static readonly IReadOnlyList<SoundPack> Catalog = Load();
    static List<SoundPack> Load()
    {
        using var stream = typeof(SoundPacks).Assembly.GetManifestResourceStream("Riff.SoundPacks.json")
            ?? throw new InvalidDataException("The sound pack catalog is missing.");
        return JsonSerializer.Deserialize<List<SoundPack>>(stream, Wire.Json)
            ?? throw new InvalidDataException("The sound pack catalog is invalid.");
    }
    // Remember catalog entries already offered, so deleted sounds stay deleted across updates.
    public static SavedState Populate(SavedState state)
    {
        var known = (state.KnownBundledSoundIds ?? []).ToHashSet();
        var existing = state.Clips.Select(c => c.Id).ToHashSet();
        var additions = Catalog.SelectMany(p => p.Sounds)
            .Where(s => !known.Contains(s.ClipId) && !existing.Contains(s.ClipId))
            .Select(s => new Clip(s.ClipId, s.Name, s.Duration)).ToList();
        known.UnionWith(Catalog.SelectMany(p => p.Sounds).Select(s => s.ClipId));
        if (additions.Count == 0 && known.SetEquals(state.KnownBundledSoundIds ?? [])) return state;
        return state with { Clips = [.. state.Clips, .. additions], KnownBundledSoundIds = known.Order().ToList(), Version = state.Version + 1 };
    }

    public static PackSound Find(string packId, string soundId) =>
        Catalog.FirstOrDefault(p => p.Id == packId)?.Sounds.FirstOrDefault(s => s.Id == soundId)
        ?? throw new ArgumentException("This sound pack is unavailable. Update both Riff apps and try again.");
}

File: Riff/Features/Sounds/SoundPacks.swift
import Foundation
import CryptoKit

struct SoundPack: Decodable, Identifiable {
    let id: String
    let name: String
    let description: String
    let icon: String
    let color: String
    let sounds: [PackSound]

}

struct PackSound: Decodable, Identifiable {
    let id: String
    let name: String
    let fileName: String
    let sourceURL: URL
    let provider: String
    let uploader: String
    let rights: String
    let byteCount: Int
    let sha256: String
    let duration: Double
    var clipID: String { "pack-" + id }

    func validate(_ data: Data) throws {
        guard data.count == byteCount,
              SHA256.hash(data: data).map({ String(format: "%02x", $0) }).joined() == sha256 else {
            throw RiffError.message("The bundled audio for \(name) is damaged. Reinstall Riff to restore it.")
        }
    }
}

enum SoundPacks {
    private static var bundle: Bundle {
#if SWIFT_PACKAGE
        Bundle.module
#else
        Bundle.main
#endif
    }

    static var clips: [Clip] { (try? load())?.flatMap { $0.sounds.map { Clip(id: $0.clipID, name: $0.name, duration: $0.duration) } } ?? [] }

    static func load() throws -> [SoundPack] {
        guard let url = bundle.url(forResource: "sound-packs", withExtension: "json") else {
            throw RiffError.message("The sound pack catalog is missing. Reinstall Riff to restore it.")
        }
        return try JSONDecoder().decode([SoundPack].self, from: Data(contentsOf: url))
    }

    static func audioURL(for sound: PackSound) throws -> URL {
        guard let url = bundle.url(forResource: sound.fileName, withExtension: nil) else {
            throw RiffError.message("The bundled audio for \(sound.name) is missing. Reinstall Riff to restore it.")
        }
        return url
    }

    static func audioData(for sound: PackSound) throws -> Data {
        let data = try Data(contentsOf: audioURL(for: sound))
        try sound.validate(data)
        return data
    }

    static func audioURL(forClipID clipID: String) -> URL? {
        guard let sound = try? load().flatMap(\.sounds).first(where: { $0.clipID == clipID }) else { return nil }
        return try? audioURL(for: sound)
    }
}
